### PR TITLE
Use ChatGPT to write an infrastructure provider

### PR DESCRIPTION
**Pull request by dkoshkin**
A fun post on how to use ChatGPT to write an infrastructure provider

### CODE REVIEW

**Review by mmcelroy91 — COMMENTED** [review]
Dang, I'm out of a job! 
image
This was definitely interesting... maybe little bot dude can help me write better "steps" when following behind code.  Definitely interested to see how the technology evolves in the future!

**Review by mmcelroy91 — APPROVED** [review]

File: content/posts/chatgpt-writes-a-capi-infrastructure-provider/index.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+title: "ChatGPT Writes a CAPI Infrastructure Provider"
+date: 2023-03-08T10:12:58-08:00
+featured: false
+tags: ["cluster api", "capi" ]
+excerpt: ADD EXCERPT HERE
+feature_image: feature.png
+---
+
+Just like many others, I first heard of [ChatGPT](https://openai.com/blog/chatgpt) at the end of 2022.
+But, besides the initial novelty of getting it to write romantic texts to send to my wife or bad jokes to send to my friends, I haven't used very much after that.
```
Comment (cmastr) [inline content/posts/chatgpt-writes-a-capi-infrastructure-provider/index.md:12]:
```suggestion
But, besides the initial novelty of getting it to write romantic texts to send to my wife or bad jokes to send to my friends, I haven't used it very much after that.
```

**Review by tillt — DISMISSED** [review]
Love it a lot -- what a great post.

I refrained from nit-picking on "I'm" etc as this post is very personal language centred and that fits the theme to me very well.

File: content/posts/chatgpt-writes-a-capi-infrastructure-provider/index.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+It even updated the reconciler code:
+
+![rename CRD](crdnamecode.png)
+
+Pretty good so far, but let's give it a hint that we actually want a Cluster API reconciler:
+
+![CAPI reconciler](capireconciler.png)
+
+If we compare to what we have so far with the [implementation contract](https://cluster-api.sigs.k8s.io/developer/providers/cluster-infrastructure.html),
+there are some required fields that are missing, let's see if ChatGTP can figure those out for us:
```
Comment (jimmidyson) [inline content/posts/chatgpt-writes-a-capi-infrastructure-provider/index.md:46]:
```suggestion
there are some required fields that are missing, let's see if ChatGPT can figure those out for us:
```

**Review by jimmidyson — CHANGES_REQUESTED** [review]

File: content/posts/chatgpt-writes-a-capi-infrastructure-provider/index.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+featured: false
+tags: ["cluster api", "capi" ]
+excerpt: "ChatGPT Writes a CAPI Infrastructure Provider"
+feature_image: feature.png
+---
+
+Just like many others, I first heard of [ChatGPT](https://openai.com/blog/chatgpt) at the end of 2022.
+But, besides the initial novelty of getting it to write romantic texts to send to my wife or bad jokes to send to my friends, I haven't used it very much after that.
+That was until I saw the headlines of Microsoft's Bing AI (powered by an even more advanced version of ChatGPT) [professing its love to a NYT tech columnist Kevin Roose](https://www.nytimes.com/2023/02/16/technology/bing-chatbot-microsoft-chatgpt.html).
+As a software engineer I got curious again and decided to find out if ChatGPT can help do my job (without it trying to break up my marriage in the process).
```
Comment (msdolbey) [inline content/posts/chatgpt-writes-a-capi-infrastructure-provider/index.md:14]:
```suggestion
As a software engineer, I got curious again and decided to find out if ChatGPT can help do my job (without it trying to break up my marriage in the process).
```

File: content/posts/chatgpt-writes-a-capi-infrastructure-provider/index.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+I decided to have ChatGPT write me a dummy Cluster API infrastructure provider.
+Sure there are great [implementation docs](https://cluster-api.sigs.k8s.io/developer/providers/implementers-guide/overview.html) 
+and even a [reference implementation for Docker](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/cluster.html?highlight=infrastruct#infrastructure-provider),
+but who has time to read all of that, especially if we can have AI do it for us!
+
+Let's give it the initial prompt "Write golang code using kubebuilder to write the cluster API infrastructure provider code"
+
+![initial prompt](initialprompt.png)
+
+We are on the right track, it knows that it needs to use [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) and something about a "Cluster".
```
Comment (msdolbey) [inline content/posts/chatgpt-writes-a-capi-infrastructure-provider/index.md:30]: Love this line

File: content/posts/chatgpt-writes-a-capi-infrastructure-provider/index.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+It even generated some (incomplete) boilerplate Go code for us. (not shown for sake of brevity)
```
Comment (msdolbey) [inline content/posts/chatgpt-writes-a-capi-infrastructure-provider/index.md:31]: Just period at the end of this line so the parenthetical statement isn't hanging.

```suggestion
It even generated some (incomplete) boilerplate Go code for us (not shown for sake of brevity).
```

File: content/posts/chatgpt-writes-a-capi-infrastructure-provider/index.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+![controlPlaneEndpoint](controlPlaneEndpoint.png)
+
+Let's prompt ChatGPT to give use one more important part of the contract, a `TemplateSpec`:
+
+![TemplateSpec](templatespec.png)
+
+### Do Something Useful
+
+At this point, ChatGPT provider a lot of boilerplate code, let's try it to use what the context it has through our interaction
```
Comment (msdolbey) [inline content/posts/chatgpt-writes-a-capi-infrastructure-provider/index.md:61]:
```suggestion
At this point, ChatGPT provided a lot of boilerplate code, let's try it to use what the context it has through our interaction
```

File: content/posts/chatgpt-writes-a-capi-infrastructure-provider/index.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+Let's see if it knows anything about the status fields:
+
+![Set status fields](setstatusfields.png)
+
+Amazing, we know have some code for our infrastructure provider that will use the information from a new Cluster object, save those values and set the proper status.
+
+### Final Thoughts
+
+At this point, ChatGPT has given us lots of useful information, sure, we can't just copy and paste the code and have it work, but it's great starting point!
```
Comment (msdolbey) [inline content/posts/chatgpt-writes-a-capi-infrastructure-provider/index.md:79]: Just a nit - I don't know if it needs to be two different sentences, but I like it this way.

```suggestion
At this point, ChatGPT has given us lots of useful information. Sure, we can't just copy and paste the code and have it work, but it's great starting point!
```

**Review by msdolbey — DISMISSED** [review]
👍 

Some small nits, but overall - great post. A little shocking and horrifying that ... I mean, I have a CKA and chatgpt is like ... challenging me here ... 

side eye at you chatgpt

**Review by tillt — APPROVED** [review]
Awesome!

**Review by cmastr — APPROVED** [review]

**Review by msdolbey — APPROVED** [review]
👍 
Meant to approve earlier, but NOW I'LL ACTUALLY DO IT, instead of leaving it as a comment like a yokel. 
Looks great

**Comment by tillt** [comment]
Ready to merge -- giving it a final glance....

**Review by jimmidyson — APPROVED** [review]

**Pull request merged**